In [1]:
import numpy as np

# Dataset XOR
X = np.array([[0,0],[0,1],[1,0],[1,1]])
y = np.array([[0],[1],[1],[0]])

# Parameter
input_size = 2
hidden_size = 2
output_size = 1
lr = 0.1

# Inisialisasi bobot
W1 = np.random.randn(input_size, hidden_size)
b1 = np.zeros((1, hidden_size))
W2 = np.random.randn(hidden_size, output_size)
b2 = np.zeros((1, output_size))

# Fungsi aktivasi
def sigmoid(x):
    return 1 / (1 + np.exp(-x))

def sigmoid_derivative(x):
    return x * (1 - x)

# Training
for epoch in range(10000):
    # Forward pass
    z1 = np.dot(X, W1) + b1
    a1 = sigmoid(z1)
    z2 = np.dot(a1, W2) + b2
    a2 = sigmoid(z2)

    # Hitung error
    error = y - a2

    # Backpropagation
    d_a2 = error * sigmoid_derivative(a2)
    d_W2 = np.dot(a1.T, d_a2)
    d_b2 = np.sum(d_a2, axis=0, keepdims=True)

    d_a1 = np.dot(d_a2, W2.T) * sigmoid_derivative(a1)
    d_W1 = np.dot(X.T, d_a1)
    d_b1 = np.sum(d_a1, axis=0, keepdims=True)

    # Update bobot
    W1 += lr * d_W1
    b1 += lr * d_b1
    W2 += lr * d_W2
    b2 += lr * d_b2

    if epoch % 1000 == 0:
        loss = np.mean(np.square(error))
        print(f"Epoch {epoch}, Loss: {loss}")

# Output akhir
print("Prediksi:")
print(a2)

Epoch 0, Loss: 0.31588712086860626
Epoch 1000, Loss: 0.24305227526805523
Epoch 2000, Loss: 0.17983983913777993
Epoch 3000, Loss: 0.06001001231356514
Epoch 4000, Loss: 0.025099336454873968
Epoch 5000, Loss: 0.014477315506819001
Epoch 6000, Loss: 0.009853172063522406
Epoch 7000, Loss: 0.007359824262360488
Epoch 8000, Loss: 0.005827322804156253
Epoch 9000, Loss: 0.004800025435994283
Prediksi:
[[0.06184057]
 [0.93151529]
 [0.93138883]
 [0.05524612]]


Tugas 1:

Ubah jumlah neuron hidden layer menjadi 3.

Bandingkan hasil loss dengan konfigurasi awal.

Tambahkan fungsi aktivasi ReLU dan bandingkan hasil.

In [2]:
import numpy as np

# Dataset XOR
X = np.array([[0,0],[0,1],[1,0],[1,1]])
y = np.array([[0],[1],[1],[0]])

# Pilih aktivasi ('sigmoid' atau 'relu')
activation_type = 'relu'   # ubah ke 'sigmoid' jika ingin membandingkan

# Parameter
input_size = 2
hidden_size = 3
output_size = 1
lr = 0.1

# Inisialisasi bobot
W1 = np.random.randn(input_size, hidden_size)
b1 = np.zeros((1, hidden_size))
W2 = np.random.randn(hidden_size, output_size)
b2 = np.zeros((1, output_size))

# Fungsi aktivasi
def sigmoid(x): return 1 / (1 + np.exp(-x))
def sigmoid_derivative(x): return x * (1 - x)

def relu(x): return np.maximum(0, x)
def relu_derivative(x): return np.where(x > 0, 1, 0)

# Pilih aktivasi
if activation_type == 'sigmoid':
    act = sigmoid; d_act = sigmoid_derivative
else:
    act = relu; d_act = relu_derivative

# Training
for epoch in range(10000):
    # Forward pass
    z1 = np.dot(X, W1) + b1
    a1 = act(z1)
    z2 = np.dot(a1, W2) + b2
    a2 = sigmoid(z2)

    # Error
    error = y - a2

    # Backpropagation
    d_a2 = error * sigmoid_derivative(a2)
    d_W2 = np.dot(a1.T, d_a2)
    d_b2 = np.sum(d_a2, axis=0, keepdims=True)

    d_a1 = np.dot(d_a2, W2.T) * d_act(a1)
    d_W1 = np.dot(X.T, d_a1)
    d_b1 = np.sum(d_a1, axis=0, keepdims=True)

    # Update
    W1 += lr * d_W1; b1 += lr * d_b1
    W2 += lr * d_W2; b2 += lr * d_b2

    if epoch % 1000 == 0:
        loss = np.mean(np.square(error))
        print(f"Epoch {epoch}, Loss: {loss}")

# Hasil
print("\nPrediksi Akhir:")
print(a2)


Epoch 0, Loss: 0.2406648259241591
Epoch 1000, Loss: 0.16711558970074314
Epoch 2000, Loss: 0.16681733880868335
Epoch 3000, Loss: 0.16675321518269734
Epoch 4000, Loss: 0.1667277969489927
Epoch 5000, Loss: 0.1667170382653612
Epoch 6000, Loss: 0.16670577356636412
Epoch 7000, Loss: 0.16670057052745652
Epoch 8000, Loss: 0.166691859641251
Epoch 9000, Loss: 0.16668857344549162

Prediksi Akhir:
[[0.33344322]
 [0.33344322]
 [0.99148434]
 [0.33344322]]


Pada percobaan pertama, jaringan menggunakan fungsi aktivasi sigmoid pada layer tersembunyinya dengan dua neuron. Hasil pelatihan menunjukkan bahwa loss turun secara signifikan hingga sekitar 0.0048 dan prediksi akhir sangat mendekati target XOR, seperti `0.06`, `0.93`, dan seterusnya. Hal ini terjadi karena sigmoid mampu memberikan perubahan nilai yang halus dan non-linear pada rentang 0 hingga 1, sehingga cocok dengan pola XOR yang memang membutuhkan pembeda non-linear untuk memisahkan kelasnya. Selain itu, dua neuron pada hidden layer sudah menjadi jumlah minimal yang cukup untuk membentuk representasi XOR, sehingga jaringan dapat belajar secara efektif tanpa mengalami kesulitan dalam pembaruan gradien.

Sementara itu, pada percobaan kedua, jaringan menggunakan ReLU dengan tiga neuron pada hidden layer. Meskipun jumlah neuronnya lebih banyak, loss berkisar pada angka 0.1666 dan tidak mengalami penurunan signifikan meskipun dilakukan training hingga 10.000 epoch. Hasil prediksi juga terlihat salah pada hampir seluruh pola XOR. Penyebabnya terletak pada sifat ReLU yang hanya aktif pada nilai positif, sehingga ketika input jaringan menghasilkan nilai negatif atau nol pada neuron tertentu, gradiennya menjadi nol dan neuron tersebut berhenti belajar. Dalam kasus XOR yang memiliki pola masukan kecil seperti `0` dan `1`, banyak aktivasi ReLU justru bernilai nol, sehingga jaringan gagal membedakan pola non-linearnya.

Dengan kata lain, meskipun ReLU sangat efektif pada jaringan besar maupun data berskala besar, pada masalah sederhana dan bernilai kecil seperti XOR, sigmoid lebih stabil dan mampu membentuk non-linearitas yang dibutuhkan. Hal tersebut menjadikan konfigurasi pertama lebih sukses dalam menyelesaikan XOR dibandingkan konfigurasi kedua meski memiliki lebih sedikit neuron.
